torch/nn/modules/batchnorm.py

```diff
@@ -316,3 +317,125 @@ def _check_input_dim(self, input):
+        affine: a boolean value that when set to ``True``, this module has
+            learnable affine parameters. Default: ``True``
+        track_running_stats: a boolean value that when set to ``True``, this
+            module tracks the running mean and variance, and when set to ``False``,
+            this module does not track such statistics and always uses batch
+            statistics in both training and eval modes. Default: ``True``
+        process_group: synchronization of stats happen within each process group
+            individually. Default behavior is synchronization across the whole
+            world
+
+    Shape:
+        - Input: :math:`(N, C, +)`
+        - Output: :math:`(N, C, +)` (same shape as input)
+
+    Examples::
+
+        >>> # With Learnable Parameters
+        >>> m = nn.SyncBatchNorm(100)
+        >>> # creating process group (optional)
+        >>> # process_ids is a list of int identifying rank ids.
+        >>> process_group = torch.distributed.new_group(process_ids)
+        >>> # Without Learnable Parameters
+        >>> m = nn.BatchNorm3d(100, affine=False, process_group=process_group)
+        >>> input = torch.randn(20, 100, 35, 45, 10)
+        >>> output = m(input)
+
+    .. _`Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift`:
+        https://arxiv.org/abs/1502.03167
+    """
+
+    def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True,
+                 track_running_stats=True, process_group=None):
+        super(SyncBatchNorm, self).__init__(num_features, eps, momentum, affine, track_running_stats)
+        self.process_group = process_group
+
+    def _check_input_dim(self, input):
+        if input.dim() <= 2:
+            raise ValueError('expected at least 3D input (got {}D input)'
+                             .format(input.dim()))
+
+    def _specify_process_group(self, process_group):
+        self.process_group = process_group
+
+    def forward(self, input):
+        # currently only GPU input is supported
+        if not input.is_cuda:
+            raise ValueError('expected input tensor to be on GPU')
+
+        self._check_input_dim(input)
+
+        exponential_average_factor = 0.0
+
+        if self.training and self.track_running_stats:
+            self.num_batches_tracked += 1
+            if self.momentum is None:  # use cumulative moving average
+                exponential_average_factor = 1.0 / self.num_batches_tracked.item()
+            else:  # use exponential moving average
+                exponential_average_factor = self.momentum
+
+        world_size = 1
+        process_group = torch.distributed.group.WORLD
+        if torch.distributed.is_initialized():
+            if self.process_group:
+                process_group = self.process_group
+            world_size = torch.distributed.get_world_size(process_group)
+
+        # fallback to framework BN when synchronization is not necessary
```

Comment:
Hmmm I'd also expect it to work from within `nn.DataParallel`, but we don't even check if it's used like that.

Reply:
nn.DataParallel is not supported in this PR yet.
DataParallel would require communication among GPUs within a single process. I'm not aware of proper upstream support for that. So we decide to upstream this single-gpu DDP sync bn first.

Reply:
I think all distributed functions you have here have been present in multi-GPU versions since 0.1 in `torch.cuda.comm`. My issue with the current implementation is that it's extremely non-composable which is why I'm skeptical about including it in the core library. Such silent assumptions about where a module can or cannot be used make it very hard for one to reason about their code.

Reply:
Basically, if you're going about adding a non-composable block, you should at least make it work only in the cases you implemented, and everything else should be a hard error. This is not the case in here.

Reply:
Agreed that I should have throw a loud error in cases not supported.
I'll add a check and error out when multi-GPU per process is detected when executing SyncBatchNorm